From fbd7cc02a96700193a7b00e5c875e8a3fb8c5e60 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 22 Apr 2026 19:09:00 +0200
Subject: [PATCH 25/47] dinput: Save and load HID joystick instance GUIDs to
 the registry.

Based on patches by Connor McAdams.

(cherry picked from commit a4a85b607006bf7a6e5c71b6d2689348546e23bf)

CW-Bug-Id: #25883
---
 dlls/dinput/joystick_hid.c    | 95 ++++++++++++++++++++++++++++++++++-
 dlls/dinput/tests/hotplug.c   |  2 +-
 dlls/dinput/tests/joystick8.c | 45 ++++++-----------
 3 files changed, 111 insertions(+), 31 deletions(-)

diff --git a/dlls/dinput/joystick_hid.c b/dlls/dinput/joystick_hid.c
index 14ffe787c68..a9c99ac0e5d 100644
--- a/dlls/dinput/joystick_hid.c
+++ b/dlls/dinput/joystick_hid.c
@@ -100,7 +100,18 @@ static HRESULT insert_cache_entry( DIDEVICEINSTANCEW *instance, const WCHAR *pat
 
 #define SWAP(x) MAKELONG( HIWORD(x), LOWORD(x) )
     LIST_FOR_EACH_ENTRY( next, &joystick_cache, struct cache_entry, entry )
+    {
         if (SWAP( next->instance.guidProduct.Data1 ) > SWAP( instance->guidProduct.Data1 )) break;
+        if (SWAP( next->instance.guidProduct.Data1 ) < SWAP( instance->guidProduct.Data1 )) continue;
+        if (*path && !*next->path)
+        {
+            instance->guidInstance = next->instance.guidInstance;
+            TRACE( "Reusing instance %s, path %s\n", debugstr_device_instance( instance ), debugstr_w( path ) );
+            next->instance = *instance;
+            wcscpy( next->path, path );
+            return S_OK;
+        }
+    }
 #undef SWAP
 
     if (FAILED(hr = cache_entry_create( instance, path, &entry ))) return hr;
@@ -110,6 +121,67 @@ static HRESULT insert_cache_entry( DIDEVICEINSTANCEW *instance, const WCHAR *pat
     return S_OK;
 }
 
+static void save_registry_instances( HKEY root )
+{
+    struct cache_entry *entry;
+    DWORD vidpid = 0, index = -1;
+    WCHAR buffer[MAX_PATH];
+
+    TRACE( "Saving cached instance GUIDs\n" );
+
+    LIST_FOR_EACH_ENTRY( entry, &joystick_cache, struct cache_entry, entry )
+    {
+        index = (entry->instance.guidProduct.Data1 == vidpid) ? index + 1 : 0;
+        vidpid = entry->instance.guidProduct.Data1;
+
+        swprintf( buffer, ARRAY_SIZE(buffer), L"VID_%04X&PID_%04X\\Calibration\\%u",
+                  LOWORD(vidpid), HIWORD(vidpid), index );
+        RegSetKeyValueW( root, buffer, L"GUID", REG_BINARY, (BYTE *)&entry->instance.guidInstance,
+                         sizeof(entry->instance.guidInstance) );
+
+        TRACE( "Saved %04x:%04x index %lu, guid %s\n", LOWORD(vidpid), HIWORD(vidpid), index,
+               debugstr_guid( &entry->instance.guidInstance ) );
+    }
+}
+
+static void load_registry_product_instances( HKEY root, DWORD vidpid )
+{
+    WCHAR name[MAX_PATH];
+
+    for (DWORD i = 0; !RegEnumKeyW( root, i, name, ARRAY_SIZE(name) ); i++)
+    {
+        DIDEVICEINSTANCEW instance = {.guidProduct = dinput_pidvid_guid };
+        DWORD len = sizeof(GUID);
+
+        instance.guidProduct.Data1 = vidpid;
+        if (RegGetValueW( root, name, L"GUID", RRF_RT_REG_BINARY, NULL, &instance.guidInstance, &len )) continue;
+        insert_cache_entry( &instance, L"" );
+
+        TRACE( "Loaded %04x:%04x index %s, guid %s\n", LOWORD(vidpid), HIWORD(vidpid), debugstr_w( name ),
+               debugstr_guid( &instance.guidInstance ) );
+    }
+}
+
+static void load_registry_instances( HKEY root )
+{
+    WCHAR name[MAX_PATH], buffer[MAX_PATH];
+    HKEY hkey;
+
+    TRACE( "Loading cached instance GUIDs\n" );
+
+    for (DWORD i = 0; !RegEnumKeyW( root, i, name, ARRAY_SIZE(name) ); i++)
+    {
+        UINT vid, pid;
+
+        if (swscanf( name, L"VID_%04X&PID_%04X", &vid, &pid ) != 2) continue;
+        swprintf( buffer, ARRAY_SIZE(buffer), L"VID_%04X&PID_%04X\\Calibration", vid, pid );
+
+        if (RegOpenKeyExW( root, buffer, 0, KEY_ENUMERATE_SUB_KEYS, &hkey )) continue;
+        load_registry_product_instances( hkey, MAKELONG(vid, pid) );
+        RegCloseKey( hkey );
+    }
+}
+
 static HRESULT get_instance_from_guid( const GUID *guid, DIDEVICEINSTANCEW *instance, WCHAR *path )
 {
     struct cache_entry *entry;
@@ -125,6 +197,7 @@ static HRESULT get_instance_from_guid( const GUID *guid, DIDEVICEINSTANCEW *inst
 
     LIST_FOR_EACH_ENTRY( entry, &joystick_cache, struct cache_entry, entry )
     {
+        if (!*entry->path) continue;
         if (IsEqualGUID( &entry->instance.guidProduct, guid )) break;
         if (IsEqualGUID( &entry->instance.guidInstance, guid )) break;
     }
@@ -1802,17 +1875,30 @@ failed:
 
 HRESULT hid_joystick_refresh_devices(void)
 {
+    static const WCHAR *dinput_path = L"System\\CurrentControlSet\\Control\\MediaProperties\\"
+                                       "PrivateProperties\\DirectInput";
+
     WCHAR *paths = NULL;
     HANDLE device;
+    HANDLE mutex;
     HRESULT hr;
+    HKEY root;
     GUID hid;
 
     TRACE( "\n" );
 
     HidD_GetHidGuid( &hid );
 
+    if (RegCreateKeyExW( HKEY_CURRENT_USER, dinput_path, 0, NULL, 0, KEY_ALL_ACCESS, NULL, &root, NULL ))
+        return DIERR_DEVICENOTREG;
+
     EnterCriticalSection( &joystick_cache_cs );
+
+    mutex = CreateMutexW( NULL, FALSE, L"__wine_dinput_reg_mutex" );
+    WaitForSingleObject( mutex, INFINITE );
+
     hid_joystick_cleanup_devices();
+    load_registry_instances( root );
 
     hr = get_device_interfaces( hid, NULL, &paths );
     for (WCHAR *path = paths; SUCCEEDED(hr) && *path; path = path + wcslen( path ) + 1)
@@ -1829,7 +1915,13 @@ HRESULT hid_joystick_refresh_devices(void)
     }
     free( paths );
 
+    save_registry_instances( root );
+    ReleaseMutex( mutex );
+    CloseHandle( mutex );
+
     LeaveCriticalSection( &joystick_cache_cs );
+    RegCloseKey( root );
+
     return hr;
 }
 
@@ -1841,7 +1933,7 @@ HRESULT hid_joystick_enum_device( DWORD type, DWORD flags, DIDEVICEINSTANCEW *in
     EnterCriticalSection( &joystick_cache_cs );
 
     LIST_FOR_EACH_ENTRY( entry, &joystick_cache, struct cache_entry, entry )
-        if (!index--) break;
+        if (*entry->path && !index--) break;
     if (&entry->entry == &joystick_cache) hr = DIERR_DEVICENOTREG;
     else
     {
@@ -2242,6 +2334,7 @@ HRESULT hid_joystick_create_device( struct dinput *dinput, const GUID *guid, IDi
                                        &impl->caps, &impl->base.instance, dinput->dwVersion );
     if (hr != DI_OK) goto failed;
 
+    impl->base.instance.guidInstance = instance.guidInstance; /* use the instance GUID from the cache */
     impl->base.caps.dwDevType = impl->base.instance.dwDevType;
     impl->is_gamepad = hid_joystick_is_gamepad(impl->base.caps.dwDevType, dinput->dwVersion);
     impl->attrs = attrs;
diff --git a/dlls/dinput/tests/hotplug.c b/dlls/dinput/tests/hotplug.c
index 10a3051e8ce..e4147797652 100644
--- a/dlls/dinput/tests/hotplug.c
+++ b/dlls/dinput/tests/hotplug.c
@@ -222,7 +222,7 @@ static BOOL test_input_lost( DWORD version )
     hr = dinput_test_create_device( version, &devinst2, &device2 );
     ok( hr == DI_OK, "Unexpected hr %#lx.\n", hr );
     ok( !!device2, "device2 is NULL.\n" );
-    todo_wine ok( !memcmp( &devinst.guidInstance, &devinst2.guidInstance, sizeof(GUID) ),
+    ok( !memcmp( &devinst.guidInstance, &devinst2.guidInstance, sizeof(GUID) ),
                   "Unexpected guidInstance.\n" );
     ref = IDirectInputDevice8_Release( device2 );
     ok( ref == 0, "Release returned %ld\n", ref );
diff --git a/dlls/dinput/tests/joystick8.c b/dlls/dinput/tests/joystick8.c
index 61577bc6538..3d7e63ceb67 100644
--- a/dlls/dinput/tests/joystick8.c
+++ b/dlls/dinput/tests/joystick8.c
@@ -6312,8 +6312,8 @@ static void test_joystick_instance_guid( DWORD version )
 
         winetest_push_context( "device %d", i );
 
-        todo_wine ok( IsEqualGUID( &expect_instances[expected_joystick - 1], &instances[i] ),
-                      "Unexpected instance %s.\n", debugstr_guid( &instances[i] ) );
+        ok( IsEqualGUID( &expect_instances[expected_joystick - 1], &instances[i] ),
+           "Unexpected instance %s.\n", debugstr_guid( &instances[i] ) );
         hr = dinput_create_device( &di, &instances[i], &device );
         ok( hr == DI_OK, "Unexpected hr %#lx.\n", hr );
 
@@ -6336,8 +6336,8 @@ static void test_joystick_instance_guid( DWORD version )
 
         winetest_push_context( "device %d", i );
 
-        todo_wine ok( IsEqualGUID( &expect_instances[expected_joystick - 1], &instances[i] ),
-                      "Unexpected instance %s.\n", debugstr_guid( &instances[i] ) );
+        ok( IsEqualGUID( &expect_instances[expected_joystick - 1], &instances[i] ),
+           "Unexpected instance %s.\n", debugstr_guid( &instances[i] ) );
         hr = dinput_create_device( &di, &instances[i], &device );
         ok( hr == DI_OK, "Unexpected hr %#lx.\n", hr );
 
@@ -6357,7 +6357,6 @@ static void test_joystick_instance_guid( DWORD version )
     {
         winetest_push_context( "device %d", i );
 
-        todo_wine_if( !(i & 0x1) )
         ok( IsEqualGUID( &expect_instances[i], &instances[i] ),
             "Unexpected guidInstance %s.\n", debugstr_guid( &instances[i] ) );
 
@@ -6398,17 +6397,13 @@ static void test_joystick_instance_guid( DWORD version )
         hid_device_start( &descs[i], 1 );
 
         hr = dinput_create_device( &di, &expect_instances[i], &device );
-        todo_wine ok( hr == DI_OK, "Unexpected hr %#lx.\n", hr );
-        if (SUCCEEDED(hr))
-        {
-            check_device_hid_serial( device, descs[i].serial_str );
-            IDirectInputDevice8_Release( device );
-        }
+        ok( hr == DI_OK, "Unexpected hr %#lx.\n", hr );
+        check_device_hid_serial( device, descs[i].serial_str );
+        IDirectInputDevice8_Release( device );
         hid_device_stop( &descs[i], 1 );
 
         hr = dinput_create_device( &di, &expect_instances[i], &device );
-        todo_wine ok( hr == E_FAIL, "Unexpected hr %#lx.\n", hr );
-        if (SUCCEEDED(hr)) IDirectInputDevice8_Release( device );
+        ok( hr == E_FAIL, "Unexpected hr %#lx.\n", hr );
 
         hid_device_start( &descs[i], 1 );
         winetest_pop_context();
@@ -6435,17 +6430,13 @@ static void test_joystick_instance_guid( DWORD version )
     hid_device_start( &descs[0], 1 );
 
     hr = dinput_create_device( &di, &expect_instances[0], &device );
-    todo_wine ok( hr == DI_OK, "Unexpected hr %#lx.\n", hr );
-    if (SUCCEEDED(hr))
-    {
-        check_device_hid_serial( device, descs[0].serial_str );
-        IDirectInputDevice8_Release( device );
-    }
+    ok( hr == DI_OK, "Unexpected hr %#lx.\n", hr );
+    check_device_hid_serial( device, descs[0].serial_str );
+    IDirectInputDevice8_Release( device );
     hid_device_stop( &descs[0], 1 );
 
     hr = dinput_create_device( &di, &expect_instances[0], &device );
-    todo_wine ok( hr == E_FAIL, "Unexpected hr %#lx.\n", hr );
-    if (SUCCEEDED(hr)) IDirectInputDevice8_Release( device );
+    ok( hr == E_FAIL, "Unexpected hr %#lx.\n", hr );
 
 
     /* Start device 1. */
@@ -6456,8 +6447,7 @@ static void test_joystick_instance_guid( DWORD version )
      * with device 0.
      */
     hr = dinput_create_device( &di, &expect_instances[0], &device );
-    todo_wine ok( hr == E_FAIL, "Unexpected hr %#lx.\n", hr );
-    if (SUCCEEDED(hr)) IDirectInputDevice8_Release( device );
+    ok( hr == E_FAIL, "Unexpected hr %#lx.\n", hr );
 
     /*
      * expect_instances[1] is not currently associated with a device, which
@@ -6469,12 +6459,9 @@ static void test_joystick_instance_guid( DWORD version )
     ok( hr == DIERR_DEVICENOTREG, "Unexpected hr %#lx.\n", hr );
 
     hr = dinput_create_device( &di, &expect_instances[0], &device );
-    todo_wine ok( hr == DI_OK, "Unexpected hr %#lx.\n", hr );
-    if (SUCCEEDED(hr))
-    {
-        check_device_hid_serial( device, descs[1].serial_str );
-        IDirectInputDevice8_Release( device );
-    }
+    ok( hr == DI_OK, "Unexpected hr %#lx.\n", hr );
+    check_device_hid_serial( device, descs[1].serial_str );
+    IDirectInputDevice8_Release( device );
     hid_device_stop( &descs[1], 1 );
 
 
-- 
2.54.0

